VIM-7242: Networking refactor

rpassis opened the pull request:
#### Ticket

[VIM-7242](https://vimean.atlassian.net/browse/VIM-7242)

#### Pull Request Checklist

- [X] Resolved any merge conflicts
- [X] No build errors or warnings are introduced
- [X] New files are written in Swift
- [X] New classes contain license headers
- [X] New classes have Documentation
- [X] New public methods have Documentation

#### Issue Summary

Currently we have multiple targets that depend directly on `AFNetworking` for various reasons. These include `VimeoKit`, `VimeoUpload`, `iOS` and `macOS` (to a lesser extent).

The goal of this PR is make the necessary changes to allow us to prevent these targets from consuming `AFNetworking` directly, and only exposing the minimum functionality required via `VimeoNetworking` as a proxy.

#### Implementation Summary

This PR has been broken down into a few different subtickets that have been previously reviewed by various members of the team. The combined changes in this feature branch include the following:

* Introduction of a new `NetworkReachabilityManager` that mirrors the API and functionality previosuly provided by `AFNetworkingReachabilityManager`. This also includes various unit testing classes.

* Small refactor of `VimeoClient` and `VimeoSessionManager` to move some of the request logic to live inside `VimeoSessionManager`.

* New `SessionManaging` & `AuthenticationListeningDelegate` protocols to abstract `VimeoClient`'s direct dependency on `VimeoSessionManager`

* Addition of various new protocols to help abstract our dependency on `AFNetworking` - `EndpointType`, `Cancellable` and `URLRequestConvertible`.

* Addition of new URL and JSON request serializers (currently unused), both thoroughly tested.

#### Reviewer Tips

To help with context and motivation for some of these changes it might be helpful to review the PRs currently open on the frameworks that currently dependend on AFNetworking:

[VimeoKit](https://github.vimeows.com/MobileApps/VimeoKit/pull/473)
[VimeoUpload](https://github.com/vimeo/VimeoUpload/pull/230)
[Vimeo iOS](https://github.vimeows.com/MobileApps/Vimeo-iOS/pull/2972)

#### How to Test

- Build and run, ensure all tests pass
- Switch to the iOS repo and checkout the branch on the PR above, ensure there are no changes in the current app behavior and that all tests pass. In particular please make sure you can upload videos as usual.

Commit:
Adds URLRequestConvertible protocol and conformances (#353)

Commit:
[VIM-7221] Replaces AFNetworking's reachability manager (#351)

* Adds NetworkReachabilityManager to project

* Refactor VimeoReachability into VimeoReachabilityMonitor

* Refactors VimeoClient to inject ReachabilityMonitoring

* Changed reachabilityChanged closure to non-optional

* Changed visibility of VimeoReachabilityMonitoring

* Changed initializer to internal

* Un-nested reachability status so it can be public typealiased

* Moved reachability status enum to separate file, added unit tests

* Major refactoring to abstract NetworkReachabilityManager and remove VimeoReachability

* Isolated properties and tests unavailable on macOS

* Renames VimeoReachabilityMonitor to VimeoReachabilityProvider

* Unit test renames, code tidy up

* Adds more documentation to VimeoReachabilityProvider

* Removes start and stop listening methods from reachability protocol

* Removes unused VimeoReachabilityMonitorTests.swift

* Renames test methods

* Cosmetic changes to spacing

* Addresses PR feedback

* Updates unit tests naming convention to follow styleguide

Commit:
VIM-7223: Replace AFNetworking request serializers (#352)

* Added HTTPMethod enum representing HTTP methods

* Adds first cut of JSON and URL encoding types

* Adds initial tests for request serializers --WIP--

* Separated request serializers into different files

* Tidy up JSONEncoding and URLEncoding tests

* Updated serializer API

* Updated VimeoRequestSerializer to use new request serializer

* Fixes bug when passed parameter is nil

* Reverts --Updated VimeoRequestSerializer to use new request serializer--

* Removes UIKit import

* Project fixes, removed duplicated protocol declaration

Commit:
[WIP] VIM-XXXX: Introduces SessionManaging protocol

* Refactors VimeoClient to consume SessionManaging protocol

* Renamed Cancellable to Cancelable, small cosmetic changes

* Small refactor of response types

* Refactors SessionManagingResponse to take a generic argument

* Renames path to uri in EndpointType

* Cosmetic changes

* Fixes extra empty line

cameo-github commented:
<!--
  0 Errors
  1 Warning: SwiftLint not installed, downl...
  1 Message: Executed 174 tests, with 0 fai...
  1 Markdown
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">SwiftLint not installed, download from https://github.com/realm/SwiftLint</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 174 tests, with 0 failures (0 unexpected) in 1.349 (1.664) seconds</td>
    </tr>
  </tbody>
</table>

See build details on [CircleCI](https://circleci.com/gh/vimeo/VimeoNetworking/1462)

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

Commit:
Extends NSError to allow for checking a 404 Not Found error (#356)

Commit:
Adds typeliases to AF security policy types (#357)

Commit:
Adds getter and setter for acceptableContentTypes

Commit:
Removes AFNetworking import from bridging header

Commit:
Fixes merge conflicts

Commit:
Adds wrapper for AFNetworking UIImageView extension

Commit:
Updates ImageLoader+Caching docs

Commit:
Adds wasCached flag on loadImage method

nicolelehrer commented on Sources/Shared/Core/SessionManaging.swift:
```diff
@@ -0,0 +1,61 @@
+    
+}
+
+/// A protocol representing an endpoint to which requests can be sent to
```
tiny nit, extra "to" at the end

nicolelehrer commented on Sources/Shared/Reachability/NetworkReachabilityManager.swift:
```diff
@@ -0,0 +1,186 @@
+    
+    /// Storage for the previous status.
+    /// TODO: Ensure read/write access to this is synchronized [RDPA, 08/29/2019]
+    private var previousStatus: NetworkReachabilityStatus?
```
Any need to follow up on this before merge?

nicolelehrer replied:
Reopening just to be sure we want to leave this as a TODO

nicolelehrer commented on Sources/Shared/VimeoClient.swift:
```diff
@@ -108,17 +93,32 @@ final public class VimeoClient {
             self.configuration = appConfiguration
             self.sessionManager = sessionManager
-            
-            VimeoReachability.beginPostingReachabilityChangeNotifications()
```
Where do we kick this off now?

nicolelehrer replied:
Ah ok, it's on init 👍 - `this reachability manager immediately starts listening
    /// for reachability changes and broadcasts them via system notification to
    /// interested subscribers.`

nicolelehrer commented on Sources/Shared/Core/JSONEncoding.swift:
```diff
@@ -0,0 +1,67 @@
+    /// The writing options to be used by the JSON serializer
+    private let options: JSONSerialization.WritingOptions
+    
+    
```
nit extra line

nicolelehrer commented on Sources/Shared/Core/URLEncoding.swift:
```diff
@@ -0,0 +1,144 @@
+}
+
+// MARK: - Private convenience utilities
+extension URLEncoding {
```
Should this be marked private in lieu of marking members?

nicolelehrer commented on Sources/Shared/Core/URLRequestConvertible.swift:
```diff
@@ -0,0 +1,30 @@
+extension URL: URLRequestConvertible {
+    func asURLRequest() throws -> URLRequest {
+        return URLRequest(url: self)
+    }
```
What are the cases when we want to obfuscate `URLRequest`?

rpassis replied:
The API I have in mind for making networking requests is something like this:

```swift
class SessionManager {
    func request(_ urlRequestConvertible: URLRequestConvertible, then callback: ...) { ... }
}
```

Having this `URLRequestConvertible` protocol and the conformances provided means we can use it in different ways:

```swift
let sessionManager = ...
let url = URL(string: ...)
sessionManager.request(url, then: { ... })
```

or

```swift
let sessionManager = ...
let urlRequest = URLRequest(url ...)
urlRequest.method = ...
sessionManager.request(urlRequest, then: { ... })
```

and also

```swift

enum UserEndpoint: EndpointType, URLRequestConvertible {

    case me
    
    var path: String { ... } 
    var method: HTTPMethod { ... }
    var parameters: Parameters? { ... }
    
    func asURLRequest() throws -> URLRequest {
         // build URLRequest here based on values defined above
    }
}

let sessionManager = ...
sessionManager.request(UserEndpoint.me, then: { ... })
```

nicolelehrer replied:
Nice - love the endpoint case

nicolelehrer commented on Sources/Shared/Core/SessionManaging.swift:
```diff
@@ -0,0 +1,61 @@
+/// on AFNetworking and the Vimeo subclasses that inherit from it,
+/// Specifically `VimeoSessionManager`, `VimeoRequestSerializer` and `VimeoResponseSerializer`
+/// The goal is to make it easier for these dependencies to be swapped out when needed.
+public protocol AuthenticationListeningDelegate {
```
Should this protocol have a more specific doc associated with it?

nicolelehrer commented on Sources/Shared/Request.swift:
```diff
@@ -47,9 +47,9 @@ public enum RetryPolicy {
+    static func defaultPolicyForMethod(for method: HTTPMethod) -> RetryPolicy {
         switch method {
-        case .GET, .DELETE, .PATCH, .POST, .PUT:
+        case .get, .delete, .patch, .post, .put, .connect, .head, .options, .trace:
```
Any chance that a method like `connect` would need to have a more forgiving retry policy?

rpassis replied:
I have no idea to be honest! But given that none of the new methods are currently used/supported, should we keep the same default for all and reconsider if/when the need to support other methods arise?

nicolelehrer commented on Sources/Shared/Core/SessionManaging.swift:
```diff
@@ -0,0 +1,61 @@
+public protocol SessionManaging {
+    
+    /// Used to invalidate the session manager
+    func invalidate()
```
To make this more flexible should we provide an input parameter of type bool that specifies if pending tasks should be aborted or not? This would let us switch on the URLSession api available

rpassis replied:
Agreed, and this was done on a [separate branch/PR](https://github.com/vimeo/VimeoNetworking/pull/359/files#diff-f2ecce9caba717801d99d9e8f3fbf565R32) along with other changes

nicolelehrer commented on Sources/Shared/VimeoSessionManager.swift:
```diff
@@ -25,62 +25,117 @@
+        case .delete:
+            return self.delete(path, parameters: parameters, success: success, failure: failure)
+        case .connect, .head, .options, .trace:
+            return nil
```
Should we add a TODO here or other explanation?

nicolelehrer reviewed (changes requested):
All tests pass 🎉 - smoke testing with client integration now but wanted to release comments sooner

Commit:
Refactors ReachabilityManaging to expose unwrapped flags using default values

Commit:
Adds assertion to unsupported HTTP methods

Commit:
Cosmetic fixes

rpassis commented:
Hey @nicolelehrer thanks again for taking a look at this PR.

I think I addressed all of your comments both here and client side so please take another look when you get a chance and let me know what you think!

Commit:
Adds documentation to AuthenticationListeningDelegate

nicolelehrer reviewed (commented):
👍 looks good @rpassis! considered approved, holding off on formal approval just so we don't accidentally merge before figuring out release strategy

Commit:
Merge branch 'develop' into feature/VIM-XXXX_AFNetworkingDeprecation

Commit:
Removes AFNetworking imports

Commit:
VIM-7293: Changes default JSON serializer behavior on VimeoSessi… (#359)

* Exposes vimeo response serializer

* Refactor request/response public API

* Documentation updates, minor API tweaks

* Changes decoders declaration to lazy:

* Removes error argument optionality

* Removes VimeoNetworking inheritance on AFNetworking

* Removes public visibility from VimeoResponseSerializer initializer

* Updates EndpointType API and conformance

* Changes URLRequestConvertible methods to public

* Adds Task protocol and extension on URLSessionTask

* Updates SessionManaging public API

* Renames VimeoNetworkingError to VNError

* Adds Request conformance to EndpointType

* Tweak to SessionManaging API

* Fixes VimeoSessionManager conformance to new SessionManaging API

* VimeoClient small refactor

* Fixes unit tests

* Exposes SessionManagingResult properties publicly

* Fixes Request URLConvertible implementation

* Small code fixes and code tidy up

* Adds data request API to SessionManaging

* Adds decodable request to API

* Adds extensions to VimeoSessionManager wrapping AFNetworking

* Small doc updates

* Changes httpSessionManager visibility to internal

* Renames VimeoClient sharedClient to shared

* Updates VimeoClient docs, structure

* Moves request constants to private String extensions

Commit:
Merge branch 'develop' into feature/VIM-XXXX_AFNetworkingDeprecation

Commit:
Merge fixes

Commit:
VIM-XXXX: VimeoResponseSerializer refactor (#369)

* Refactors VimeoResponseSerializer
* Marks unused methods in response serializer as private
* Updates documentation and method signature
* Small tweaks
* Updates documentation

Commit:
VIM-XXXX: Removes shared client from networking library (#370)

* Removes shared client from networking library
* Cosmetic changes

Commit:
VIM-XXXX: Adds decodable requests to VimeoClient (#371)

* Adds decodable request to VimeoClient
* Adds resume method to RequestToken
* Fixes spelling, swaps order of function args

Commit:
VIM-XXXX: VimeoRequestSerializer refactor (#368)

* Refactors VimeoRequestSerializer
* Code formatting changes
* Guards against nil requests, extract shortVersionString into extension

Commit:
Adds TODO comment on VimeoClient

Commit:
Adds ability to configure a VimeoClient instance

Commit:
Small refactor, moving files around

Commit:
Removes inout parameter from VimeoClient(configure:) method

Commit:
Merge remote-tracking branch 'origin/develop' into feature/VIM-XXXX_AFNetworkingDeprecation

Commit:
Fixes HTTPMethod casing

nicolelehrer commented on Sources/Shared/Reachability/NetworkReachabilityStatus.swift:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    var isCellular: Bool {
+        #if os(iOS) || os(tvOS)
```
We should support watchOS here if we're mentioning it on reachability tests

rpassis replied:
Removed from tests - this is not supported on watchOS

rpassis replied:
I also added a copy of the `Mutex` type we currently have on `VimeoKit` here to ensure we are safely mutating `previousStatus` on the reachability manager.

nicolelehrer commented on Tests/Shared/Reachability/NetworkingReachabilityManagerTests.swift:
```diff
@@ -0,0 +1,161 @@
+        XCTAssertNil(manager)
+    }
+
+    func testThatAddressManagerCanBeDeinitialized() {
```
Should this be `testThatDefaultManager...`?

nicolelehrer commented on Sources/Shared/VimeoResponseSerializer.swift:
```diff
@@ -26,24 +26,49 @@
+    // The response serializer used to serialize data responses
+    private let jsonResponseSerializer: AFJSONResponseSerializer
+
+    init(jsonResponseSerializer: AFJSONResponseSerializer = AFJSONResponseSerializer()) {
```
Nit - documentation

nicolelehrer commented on Sources/Shared/VimeoClient.swift:
```diff
@@ -347,18 +448,31 @@ final public class VimeoClient {
+        guard error.code != NSURLErrorCancelled else {
+            // TODO: This error never gets propagated up the chain because we don't call the completion closure here.
+            // We need to investigate whether adding the callback here will cause any unforeseen side effects on calling
+            // sites before fixing it. [RDPA 10/16/2019]
```
Good call to investigate due to cases like backing out of a request by leaving a page, upload, etc

nicolelehrer commented on Tests/Factory/VimeoClientFactory.swift:
```diff
@@ -0,0 +1,41 @@
+import Foundation
+@testable import VimeoNetworking
+
+func makeVimeoClient(
```
nit - docs for public API

nicolelehrer commented on Tests/Factory/VimeoClientFactory.swift:
```diff
@@ -0,0 +1,41 @@
+}
+
+private extension AppConfiguration {
+    static let fake: AppConfiguration = {
```
Could we potentially isolate `fake` to just the stub `clientIdentifier` and `clientSecret`, since the rest of the configuration is valid? Maybe break out all constants to default params? For example, we would perhaps want it to be apparent what the default API version is.

rpassis replied:
Not sure if I follow this here - also note that this is for the unit test target only so it is definitely a `fake` in this case.

nicolelehrer replied:
Ah now I see this is only for unit tests 👍

nicolelehrer commented on Tests/Factory/VimeoClientFactory.swift:
```diff
@@ -0,0 +1,41 @@
+    static let fake: AppConfiguration = {
+        return AppConfiguration(
+            clientIdentifier: "{CLIENT_ID}",
+            clientSecret: "{CLIENT_SECRET}",
```
Should we consider a compiler directive? like #warning("Add your client secret") or too annoying. Perhaps bundling these into a client struct or tuple?

rpassis replied:
This is only used in the test target and we don't require the correct ID and secret to be set. I will just rename it FAKE_ID / FAKE_SECRET to make it clear that this ok.

nicolelehrer commented on Sources/Shared/Core/SessionManaging.swift:
```diff
@@ -0,0 +1,71 @@
+import Foundation
+
+public struct SessionManagingResult<T> {
+    public let request: URLRequest?
```
Should we consider enforcing passing the request? No strong preference on my end, just a thought

rpassis replied:
That would be preferrable for sure but unfortunately at this point we cannot guarantee that a request will exist - we may pass in a `URLRequestConvertible` type which attempts to create a `URLRequest` and throws an error if it fails.

nicolelehrer commented on Sources/Shared/VimeoSessionManager.swift:
```diff
@@ -26,59 +26,419 @@
-    // MARK: - Authentication
-    
+
+    public func invalidate(cancelingPendingTasks cancelPendingTasks: Bool) {
```
nit - Docs for public API?

nicolelehrer commented on Sources/Shared/VimeoSessionManager.swift:
```diff
@@ -26,59 +26,419 @@
+    ///   Note that this API makes no guarantees as to which queue the callback will be executed on.
+    ///
+    /// - Returns: the `Task` object related to the request.
+    public func request<T: Decodable>(
```
🔥 🔥 🔥

nicolelehrer commented on Sources/Shared/VimeoSessionManager.swift:
```diff
@@ -26,59 +26,419 @@
+    ///   Note that this API makes no guarantees as to which queue the callback will be executed on.
+    ///
+    /// - Returns: the `Task` object related to the request.
+    public func upload(
```
Could we make it explicit that this is for background capable uploads in the documentation?

rpassis replied:
As far as I am aware there is nothing in this method that restricts it (or makes it specifically suitable) for background uploads. Isn't this the domain of the upload library?

nicolelehrer replied:
This method will return an instance of https://developer.apple.com/documentation/foundation/nsurlsessionuploadtask?language=objc which is background capable. So it's possible to support background uploads but it isn't mandated (and depends heavily on the session being configured for background tasks) so on second thought I think you're right in that it's not necessary.

nicolelehrer commented on Sources/Shared/Core/VNError.swift:
```diff
@@ -0,0 +1,24 @@
+//
+//  VimeoNetworkingError.swift
```
Do we have to keep the file name short due to linting?

rpassis replied:
I renamed the class and forgot to rename the file, fixed!

nicolelehrer commented on Sources/Shared/Core/AuthenticatingListeningDelegate.swift:
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+/// A type that listens to and responds to authentication status changes
+public protocol AuthenticationListeningDelegate {
```
`AuthenticationListening` or `AuthenticationListener`? or does delegate help infer something else

rpassis replied:
Yes good call! At first I wanted to separate SessionManaging and AuthenticationListening as separate types because it felt they are responsible for very different things however after looking at this again I think it makes more sense to simplify this to one type - `SessionManaging` that inherits from `AuthenticationListener`.

nicolelehrer commented on Sources/Shared/Core/Request+EndpointType.swift:
```diff
@@ -0,0 +1,32 @@
+            throw VimeoNetworkingError.invalidURL
+        }
+        var urlRequest = URLRequest(url: url)
+        headers?.forEach { field, value in
```
nit - explicit self

nicolelehrer commented on Sources/Shared/Core/Request+EndpointType.swift:
```diff
@@ -0,0 +1,32 @@
+    // TODO: This method should be able to serialize its own parameters however with the current
+    // way the SessionManager is designed we are unable to do that as it would require pretty
+    // substantial refactoring. So for now the request serialization is the responsibility of the
+    // SessionManaging type [RDPA 10/02/2019].
```
Note to check in on this part

rpassis replied:
I'd like to leave that TODO there for the future as we are unable to fix this with the way things are currently structured.

nicolelehrer reviewed (changes requested):
Looking good! Tests pass 👍 Minor things to address

Commit:
Removes conditional check for watchOS

Commit:
Tweak to VimeoResponseSerializer initializer, adds documentation

Commit:
Renames test

Commit:
Adds documentation for VimeoClient factory in Test target

Commit:
Replaces dummy client id and secret with explicit fake naming

Commit:
Adds documentation for invalidate(cancelPendingTasks)

Commit:
Renames VNError to VimeoNetworkingError

Commit:
Refactors SessionManaging and AuthenticationListening composition

Commit:
Adds explicit self

Commit:
Adds lock/unlock methods for thread safety when mutating state of reachability manager

rpassis commented:
Thanks for all the feedback Nicole! I think I addressed almost all of your questions and comments, and left a couple of responses for you to take a look.

Commit:
Small tweaks

Commit:
Updates VimeoRequestSerializer initializer

Commit:
Revert "Updates VimeoRequestSerializer initializer"

This reverts commit 7e219df47d0ace099d4f5072887ffa44df810f76.

nicolelehrer reviewed (approved):
Just need to sync with dev then good to go! 🎉

Commit:
Merge remote-tracking branch 'origin/develop' into feature/VIM-XXXX_AFNetworkingDeprecation

Commit:
Removes unnecessary file

rpassis commented:
Thanks @nicolelehrer!
